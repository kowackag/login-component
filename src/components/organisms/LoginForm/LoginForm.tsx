import { InputField } from "@/components/molecules/InputField/InputField";
import { Checkbox } from "@/components/molecules/Checkbox/Checkbox";
import { Button } from "@/components/atoms/Button/Button";

import { LoginFormType } from "@/pages/LoginPage/LoginPage";

export const LoginForm = ({ onSubmit, errors, register }: LoginFormType) => {
	const {
		name: emailInputName,
		onChange: onEmailInputChange,
		onBlur: onEmailInputBlur,
		ref: emailInputRef,
	} = register("email");
	const {
		name: passwordInputName,
		onChange: onPasswordInputChange,
		onBlur: onPasswordInputBlur,
		ref: passwordInputRef,
	} = register("password");
	const {
		name: checkboxName,
		onChange: onCheckboxChange,
		onBlur: onCheckboxBlur,
		ref: checkboxRef,
	} = register("toSave");
	return (
		<form onSubmit={onSubmit}>
			<InputField
				name={emailInputName}
				label="Email"
				onChange={onEmailInputChange}
				onBlur={onEmailInputBlur}
				inputRef={emailInputRef}
				error={errors?.email?.message}
			/>
			<InputField
				name={passwordInputName}
				label="Password"
				type="password"
				onChange={onPasswordInputChange}
				onBlur={onPasswordInputBlur}
				inputRef={passwordInputRef}
				error={errors?.password?.message}
			/>
			<Checkbox
				name={checkboxName}
				label="Remember me?"
				onChange={onCheckboxChange}
				onBlur={onCheckboxBlur}
				inputRef={checkboxRef}
				error={errors?.email?.message}
			/>
			<Button text="Login" type="submit" />
		</form>
	);
};
