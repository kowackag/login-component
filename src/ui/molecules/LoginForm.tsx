import { useForm } from "react-hook-form";
import { zodResolver } from "@hookform/resolvers/zod";

import { schema } from "@/views/authSchema";
import { ConfirmButton } from "@ui/atoms/ConfirmButton";

import { FormDataType } from "@/utils/shared.types";
import { FormField } from "@ui/molecules/FormField";

export const LoginForm = () => {
	const {
		register,
		formState: { errors },
		handleSubmit,
	} = useForm<Zod.output<typeof schema>>({
		resolver: zodResolver(schema),
		defaultValues: {
			email: "",
			password: "",
		},
	});
	const onSubmit = (data: FormDataType) => console.log(data);
	return (
		<form onSubmit={handleSubmit(onSubmit)}>
			<FormField name="email" label="Email" register={register} error={errors.email?.message} />
			<FormField
				name="password"
				label="Password"
				type="password"
				register={register}
				error={errors.password?.message}
			/>
		
			<ConfirmButton>Login</ConfirmButton>
		</form>
	);
};
