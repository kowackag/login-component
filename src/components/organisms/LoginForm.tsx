import { useForm } from "react-hook-form";
import { zodResolver } from "@hookform/resolvers/zod";

import { schema } from "@/pages/authSchema";
import { ConfirmButton } from "@/components/atoms/ConfirmButton";

import { FormDataType } from "@/utils/shared.types";
import { FormField } from "@/components/molecules/FormField";
import { Checkbox } from "../molecules/Checkbox";

export const LoginForm = () => {
	const {
		register,
		formState: { errors },
		handleSubmit,
	} = useForm<Zod.output<typeof schema>>({
		resolver: zodResolver(schema),
		defaultValues: {
			email: "",
			password: "",
			saved: false,
		},
	});
	const onSubmit = (data: FormDataType) => console.log(data);
	return (
		<form onSubmit={handleSubmit(onSubmit)}>
			<FormField name="email" label="Email" register={register} error={errors.email?.message} />
			<FormField
				name="password"
				label="Password"
				type="password"
				register={register}
				error={errors.password?.message}
			/>
			<Checkbox name="saved" label="Remember me?" register={register} />
			<ConfirmButton>Login</ConfirmButton>
		</form>
	);
};
