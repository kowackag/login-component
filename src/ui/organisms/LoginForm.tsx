import { useForm } from "react-hook-form";
import { zodResolver } from "@hookform/resolvers/zod";

import { schema } from "@/views/authSchema";
import { ConfirmButton } from "@ui/atoms/ConfirmButton";

import { FormDataType } from "@/utils/shared.types";
import { FormField } from "@ui/molecules/FormField";
import { Checkbox } from "../molecules/Checkbox";

export const LoginForm = () => {
	const {
		register,
		formState: { errors },
		handleSubmit,
	} = useForm<Zod.output<typeof schema>>({
		resolver: zodResolver(schema),
		defaultValues: {
			email: "",
			password: "",
			saved: false,
		},
	});
	const onSubmit = (data: FormDataType) => console.log(data);
	return (
		<form onSubmit={handleSubmit(onSubmit)}>
			{/* CredentialsFields */}
			<FormField
				name="email"
				label="Email"
				// type="email"
				register={register}
				error={errors.email?.message}
			/>
			<FormField
				name="password"
				label="Password"
				type="password"
				register={register}
				error={errors.password?.message}
			/>
			<Checkbox name="saved" label="Remember me?" register={register} />
			<ConfirmButton>Login</ConfirmButton>
		</form>
	);
};
