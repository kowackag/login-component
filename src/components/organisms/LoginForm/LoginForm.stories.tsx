import type { Meta } from "@storybook/react";
import { useForm } from "react-hook-form";
import { fn } from "@storybook/test";

import { LoginForm } from "./LoginForm";
import { LoginFormDataTypes } from "@/pages/LoginPage/LoginPage";

const meta: Meta<typeof LoginForm> = {
	component: LoginForm,
	decorators: [
		(Story) => (
			<div style={{ width: "360px" }}>
				<Story />
			</div>
		),
	],
};

export default meta;

const onSubmit = fn();
const errors = {
	email: { message: "it is required", type: "required" },
	password: { message: "some error", type: "required" },
};

export const Default = () => {
	const { register } = useForm<LoginFormDataTypes>();

	return <LoginForm onSubmit={onSubmit} register={register} errors={errors} />;
};
