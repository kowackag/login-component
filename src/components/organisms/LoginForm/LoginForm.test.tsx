import { expect, it, describe, vi } from "vitest";
import { fireEvent, render, screen } from "@testing-library/react";
import { LoginForm } from "./LoginForm";
import { FieldErrors, UseFormRegister, FieldValues } from "react-hook-form";
type LoginFieldsTypes = { email: string; password: string; toSave: boolean };

describe("LoginForm", () => {
	it("LoginForm exists", async () => {
		const mockFn = vi.fn();
		const register: UseFormRegister<LoginFieldsTypes> = vi.fn((name) => {
			return {
				name: name as FieldValues["name"],
				onChange: mockFn,
				onBlur: mockFn,
				ref: mockFn,
			};
		});
		render(<LoginForm onSubmit={mockFn} register={register} />);

		const form = screen.getByTestId("login-form");
		expect(form).toBeInTheDocument();
	});

	it("has appropriate numbers of children", async () => {
		const mockFn = vi.fn();
		const register: UseFormRegister<LoginFieldsTypes> = vi.fn((name) => {
			return {
				name: name as FieldValues["name"],
				onChange: mockFn,
				onBlur: mockFn,
				ref: mockFn,
			};
		});
		render(<LoginForm onSubmit={mockFn} register={register} />);
		const form = screen.getByTestId("login-form");
		expect(form.children.length).toBe(4);
	});

	it("shows appropriate error numbers and errors message", async () => {
		const mockFn = vi.fn();
		const register: UseFormRegister<LoginFieldsTypes> = vi.fn((name) => {
			return {
				name: name as FieldValues["name"],
				onChange: mockFn,
				onBlur: mockFn,
				ref: mockFn,
			};
		});
		const errors: FieldErrors<LoginFieldsTypes> = {
			email: {
				message: "Invalid email",
				type: "invalid form",
			},
		};
		render(<LoginForm onSubmit={mockFn} register={register} errors={errors} />);
		const errorList = screen.getAllByTestId("error");
		expect(errorList).toHaveLength(1);
		expect(errorList[0]).toHaveTextContent("Invalid email");
	});
	it("has toSave checkbox that isunchecked by default ", async () => {
		const mockFn = vi.fn();
		const register: UseFormRegister<LoginFieldsTypes> = vi.fn((name) => {
			return {
				name: name as FieldValues["name"],
				onChange: mockFn,
				onBlur: mockFn,
				ref: mockFn,
			};
		});
		const { container } = render(<LoginForm onSubmit={mockFn} register={register} />);

		const checkbox = container.querySelector("input[data-testid='checkbox']") as HTMLInputElement;
		expect(checkbox).toBeInTheDocument();
		expect(checkbox.checked).toBe(false);
		expect(checkbox).toHaveAttribute("type", "checkbox");
		fireEvent.click(checkbox);
		expect(mockFn).toBeCalled();
		expect(checkbox.checked).toBe(true);
	});

	it("has button type submit", async () => {
		const mockFn = vi.fn();
		const register: UseFormRegister<LoginFieldsTypes> = vi.fn((name) => {
			return {
				name: name as FieldValues["name"],
				onChange: mockFn,
				onBlur: mockFn,
				ref: mockFn,
			};
		});
		render(<LoginForm onSubmit={mockFn} register={register} />);

		const button = screen.getByRole("button", { name: /login/i });
		expect(button).toHaveAttribute("type", "submit");
	});

	it("renders correctly", () => {
		const mockFn = vi.fn();
		const register: UseFormRegister<LoginFieldsTypes> = vi.fn((name) => {
			return {
				name: name as FieldValues["name"],
				onChange: mockFn,
				onBlur: mockFn,
				ref: mockFn,
			};
		});
		const { container } = render(<LoginForm onSubmit={mockFn} register={register} />);
		expect(container).toMatchSnapshot();
	});
});
