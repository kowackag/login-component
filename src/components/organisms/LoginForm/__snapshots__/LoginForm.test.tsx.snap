// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginForm > renders correctly 1`] = `
<div>
  <form
    data-testid="login-form"
  >
    <div
      class="relative py-3"
    >
      <label
        class="inline-block font-primaryMedium"
        for="email"
      >
        Email
      </label>
      <input
        class="w-full rounded-md border-2 border-neutral-300 bg-transparent px-2 leading-9 outline-none"
        id="email"
        name="email"
        type="text"
      />
    </div>
    <div
      class="relative py-3"
    >
      <label
        class="inline-block font-primaryMedium"
        for="password"
      >
        Password
      </label>
      <input
        class="w-full rounded-md border-2 border-neutral-300 bg-transparent px-2 leading-9 outline-none"
        id="password"
        name="password"
        type="password"
      />
    </div>
    <div
      class="justify-beetwen mb-4 flex align-middle"
    >
      <input
        class="peer fixed h-5 w-5 opacity-0"
        data-testid="checkbox"
        id="toSave"
        name="toSave"
        type="checkbox"
      />
      <p
        class="mr-3 inline-block h-5 w-5 rounded-sm border border-neutral-300 p-0 leading-5 text-white peer-checked:border-none peer-checked:bg-bright-rose "
      >
        <svg
          height="16px"
          width="16px"
        >
          <use
            href="/icons.svg#check"
          />
        </svg>
      </p>
      <label
        class="font-primaryMedium leading-5"
        for="toSave"
      >
        Remember me?
      </label>
    </div>
    <button
      class="my-2 w-full rounded-lg bg-bright-rose text-center font-primaryMedium text-sm uppercase leading-10 tracking-widest text-neutral-100 shadow-btn"
      data-testid="btn-submit"
      type="submit"
    >
      Login
    </button>
  </form>
</div>
`;
